pkg/controller/history/controller_history.go

```diff
@@ -0,0 +1,490 @@
+}
+
+// NewHistory returns an instance of Interface that uses client to communicate with the API Server and lister to list
+// ControllerRevisions. This method should be used to create an Interface for all scenarios other than testing.
+func NewHistory(client clientset.Interface, lister appslisters.ControllerRevisionLister) Interface {
+	return &realHistory{client, lister}
+}
+
+// NewFakeHistory returns an instance of Interface that uses informer to create, update, list, and delete
+// ControllerRevisions. This method should be used to create an Interface for testing purposes.
+func NewFakeHistory(informer appsinformers.ControllerRevisionInformer) Interface {
+	return &fakeHistory{informer.Informer().GetIndexer(), informer.Lister()}
+}
+
+type realHistory struct {
+	client clientset.Interface
+	lister appslisters.ControllerRevisionLister
+}
+
+func (rh *realHistory) ListControllerRevisions(parent metav1.Object, selector labels.Selector) ([]*apps.ControllerRevision, error) {
+	// List all revisions in the namespace that match the selector
+	history, err := rh.lister.ControllerRevisions(parent.GetNamespace()).List(selector)
+	if err != nil {
+		return nil, err
+	}
+	var owned []*apps.ControllerRevision
+	for i := range history {
+		ref := controller.GetControllerOf(history[i])
+		if ref == nil || ref.UID == parent.GetUID() {
+			owned = append(owned, history[i])
+		}
+
+	}
+	return owned, err
+}
+
+func (rh *realHistory) CreateControllerRevision(parent metav1.Object, revision *apps.ControllerRevision) (*apps.ControllerRevision, error) {
+	// Initialize the probe to 0
+	probe := uint32(0)
+
+	// Clone the input
+	any, err := scheme.Scheme.DeepCopy(revision)
+	if err != nil {
+		return nil, err
+	}
+	clone := any.(*apps.ControllerRevision)
+
+	// Continue to attempt to create the revision updating the name with a new hash on each iteration
+	for {
+		var hash uint32
+		// The first attempt uses no probe to resolve collisions
+		if probe > 0 {
+			hash = HashControllerRevision(revision, &probe)
+		} else {
+			hash = HashControllerRevision(revision, nil)
+		}
+		// Update the revisions name and labels
+		clone.Name = ControllerRevisionName(parent.GetName(), hash)
+		created, err := rh.client.Apps().ControllerRevisions(parent.GetNamespace()).Create(clone)
+		if errors.IsAlreadyExists(err) {
+			probe++
+			continue
+		}
+		return created, err
+	}
+}
+
+func (rh *realHistory) UpdateControllerRevision(revision *apps.ControllerRevision, newRevision int64) (*apps.ControllerRevision, error) {
+	obj, err := scheme.Scheme.DeepCopy(revision)
+	if err != nil {
+		return nil, err
+	}
+	clone := obj.(*apps.ControllerRevision)
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if clone.Revision == newRevision {
+			return nil
+		}
+		clone.Revision = newRevision
+		updated, updateErr := rh.client.Apps().ControllerRevisions(clone.Namespace).Update(clone)
```

Comment:
I am planning on updating statefulset and history to use AppsV1beta1 in a follow up PR. I think it is best to update them all, simultaneously, with a PR to do explicitly that, but I'd be willing to update history now if there is a strong reason.